File: src/main/java/me/deecaad/weaponmechanicsplus/weapon/modifiers/util/Operation.java
package me.deecaad.weaponmechanicsplus.weapon.modifiers.util;

/**
 * Outlines a simple math operation, like add/multiply.
 */
public enum Operation {

    ADD {
        @Override
        public double evaluate(double a, double b) {
            return a + b;
        }

        @Override
        public int evaluate(int a, int b) {
            return a + b;
        }
    }, MULTIPLY {
        @Override
        public double evaluate(double a, double b) {
            return a * b;
        }

        @Override
        public int evaluate(int a, int b) {
            return a * b;
        }
    }, SET {
        @Override
        public double evaluate(double a, double b) {
            return b;
        }

        @Override
        public int evaluate(int a, int b) {
            return b;
        }
    };

    public abstract double evaluate(double a, double b);

    public abstract int evaluate(int a, int b);
}

File: src/main/kotlin/com/cjcrafter/weaponmechanicsplus/weapon/modifiers/util/Operation.kt
/*
 * Copyright (c) 2026. All rights reserved. Distribution of this file, similar
 * files, related files, or related projects is strictly controlled.
 */

package com.cjcrafter.weaponmechanicsplus.weapon.modifiers.util

/**
 * Outlines a simple math operation, like add/multiply.
 */
enum class Operation {
    ADD {
        override fun evaluate(a: Double, b: Double): Double {
            return a + b
        }

        override fun evaluate(a: Float, b: Float): Float {
            return a + b
        }

        override fun evaluate(a: Int, b: Int): Int {
            return a + b
        }
    },
    MULTIPLY {
        override fun evaluate(a: Double, b: Double): Double {
            return a * b
        }

        override fun evaluate(a: Float, b: Float): Float {
            return a * b
        }

        override fun evaluate(a: Int, b: Int): Int {
            return a * b
        }
    },
    SET {
        override fun evaluate(a: Double, b: Double): Double {
            return b
        }

        override fun evaluate(a: Float, b: Float): Float {
            return b
        }

        override fun evaluate(a: Int, b: Int): Int {
            return b
        }
    };

    abstract fun evaluate(a: Double, b: Double): Double
    abstract fun evaluate(a: Float, b: Float): Float
    abstract fun evaluate(a: Int, b: Int): Int
}